trieste/acquisition/function.py

```diff
@@ -528,15 +509,134 @@ def prepare_acquisition_function(
+    observation_shape: ShapeLike,
+) -> Callable[[QueryPoints], tf.Tensor]:
+    """
+    Returns independent samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    N is the number of query points and L is the number of latent processes.
+    :param dataset:
+    :param model:
+    :param sample_size:
+    :param observation_shape:
+    :return: a [S, N, L] tensor
+    """
+    eps = tf.random.normal(
+        [sample_size] + list(observation_shape), dtype=dataset.observations.dtype
+    )
+
+    def sample(at: tf.Tensor) -> tf.Tensor:
+        tf.debugging.assert_shapes([at, "ND"])
+        mean, cov = model.predict(at)  # [N, L], [N, L]
+        return mean + tf.math.sqrt(cov) * eps[:, None, :]  # [S, N, L]
+
+    return sample
+
+
+# todo the batch size problem
+def batch_latent_sampler_with_reparametrisation(
+    dataset: Dataset,
+    model: ProbabilisticModel,
+    sample_size: int,
+    batch_size: int,
+    jitter: Union[tf.Tensor, float] = DEFAULTS.JITTER,
+) -> Callable[[BatchQueryPoints], tf.Tensor]:
+    """
+    Returns batch-correlated samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    the batch size B must be compatible between eps and at,
+    N is the number of batches of query points
+    and L is the number of latent processes.
+    :param dataset:
+    :param model:
+    :param sample_size:
+    :param batch_size:
+    :param jitter:
+    :return: a [S, N, B, L] tensor
+    """
+    eps = tf.random.normal(
+        [dataset.observations.shape[-1:], batch_size, sample_size], dtype=dataset.observations.dtype
+    )  # [L, B, S]
+
+    def sample(at: tf.Tensor) -> tf.Tensor:
+        tf.debugging.assert_shapes([(at, [..., sample_size, batch_size, "_"])])
+
+        # todo check this is the same as what victor wrote
+        mean, cov = model.predict_full_cov(at)  # [N, B, L], [N, L, B, B]
+        identity = tf.eye(batch_size, dtype=cov.dtype)  # [B, B]
+        cov_cholesky = tf.linalg.cholesky(cov + jitter * identity)  # [N, L, B, B]
+        return mean + tf.transpose(cov_cholesky @ eps, [3, 0, 2, 1])  # [S, N, B, L]
+
+    return sample
+
+
+class BatchReparametrizationSampler:
+    def __init__(
+        self,
+        sample_size: int,
+        dataset: Dataset,
+        model: ProbabilisticModel,
+        jitter: float = DEFAULTS.JITTER
+    ):
+        self._model = model
+        self._target_dimension = dataset.observations.shape[-1:]
+        self._sample_size = sample_size
+        self._eps = tf.Variable(
+            tf.zeros([self._target_dimension, 0, sample_size], dtype=dataset.observations.dtype),
+            shape=[self._target_dimension, None, sample_size],
+        )
+        self._jitter = jitter
+
+    def sample(self, at: tf.Tensor) -> tf.Tensor:
+        mean, cov = self._model.predict_full_cov(at)  # [N, B, L], [N, L, B, B]
+
+        batch_size = mean.shape[-1]
+
+        if tf.size(self._eps) == 0:
```

Comment:
todo this logic assumes target_dimension, batch_size and sample_size are all non-zero, which is a reasonable assumption, but we should add error checking for it